bot.go

```diff
@@ -18,33 +18,43 @@ import (
 	"github.com/technoweenie/multipartstreamer"
 )
 
+// BotAPI configuration object to allow custom setups.
+//
+// BotAPIConfig allows to set http.Client and http.Server
+type BotAPIConfig struct {
+	Client   *http.Client
+	Server   *http.Server
+	ServeMux *http.ServeMux
+}
+
 // BotAPI allows you to interact with the Telegram Bot API.
 type BotAPI struct {
 	Token  string `json:"token"`
 	Debug  bool   `json:"debug"`
 	Buffer int    `json:"buffer"`
 
-	Self            User         `json:"-"`
-	Client          *http.Client `json:"-"`
+	Self            User          `json:"-"`
+	Config          *BotAPIConfig `json:"-"`
 	shutdownChannel chan interface{}
 }
 
 // NewBotAPI creates a new BotAPI instance.
 //
 // It requires a token, provided by @BotFather on Telegram.
 func NewBotAPI(token string) (*BotAPI, error) {
-	return NewBotAPIWithClient(token, &http.Client{})
+	return NewBotAPIWithConfig(token, &BotAPIConfig{http.DefaultClient,
+		nil, http.DefaultServeMux})
 }
 
 // NewBotAPIWithClient creates a new BotAPI instance
-// and allows you to pass a http.Client.
+// and allows you to pass BotAPIConfig.
 //
 // It requires a token, provided by @BotFather on Telegram.
-func NewBotAPIWithClient(token string, client *http.Client) (*BotAPI, error) {
+func NewBotAPIWithConfig(token string, config *BotAPIConfig) (*BotAPI, error) {
```

Comment:
Instead of replacing this method, you can leave the existing `NewBotAPIWithClient` and add a new `NewBotAPIWithConfig`. 

The `WithClient` variant can then construct a `Config` struct and pass it to `WithConfig`. Everything else looks good!